aws/rust-runtime/aws-runtime/src/auth.rs

```diff
@@ -3,5 +3,227 @@
+    };
+    settings.expires_in = operation_config.signing_options.expires_in;
+    settings
+}
+
+#[derive(Debug)]
+enum SigV4SigningError {
+    MissingOperationSigningConfig,
+    MissingSigningRegion,
+    #[cfg(feature = "sigv4a")]
+    MissingSigningRegionSet,
+    MissingSigningName,
+    WrongIdentityType(Identity),
+    BadTypeInEndpointAuthSchemeConfig(&'static str),
+}
+
+impl fmt::Display for SigV4SigningError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use SigV4SigningError::*;
+        let mut w = |s| f.write_str(s);
+        match self {
+            MissingOperationSigningConfig => w("missing operation signing config"),
+            MissingSigningRegion => w("missing signing region"),
+            #[cfg(feature = "sigv4a")]
+            MissingSigningRegionSet => w("missing signing region set"),
+            MissingSigningName => w("missing signing name"),
+            WrongIdentityType(identity) => {
+                write!(f, "wrong identity type for SigV4/sigV4a. Expected AWS credentials but got `{identity:?}`")
+            }
+            BadTypeInEndpointAuthSchemeConfig(field_name) => {
+                write!(
+                    f,
+                    "unexpected type for `{field_name}` in endpoint auth scheme config",
+                )
+            }
+        }
+    }
+}
+
+impl StdError for SigV4SigningError {
+    fn source(&self) -> Option<&(dyn StdError + 'static)> {
+        match self {
+            Self::MissingOperationSigningConfig => None,
+            Self::MissingSigningRegion => None,
+            #[cfg(feature = "sigv4a")]
+            Self::MissingSigningRegionSet => None,
+            Self::MissingSigningName => None,
+            Self::WrongIdentityType(_) => None,
+            Self::BadTypeInEndpointAuthSchemeConfig(_) => None,
+        }
+    }
+}
+
+fn extract_endpoint_auth_scheme_signing_name(
+    endpoint_config: &AuthSchemeEndpointConfig<'_>,
+) -> Result<Option<SigningName>, SigV4SigningError> {
+    use SigV4SigningError::BadTypeInEndpointAuthSchemeConfig as UnexpectedType;
+
+    match extract_field_from_endpoint_config("signingName", endpoint_config) {
+        Some(Document::String(s)) => Ok(Some(SigningName::from(s.to_string()))),
+        None => Ok(None),
+        _ => Err(UnexpectedType("signingName")),
+    }
+}
+
+fn extract_endpoint_auth_scheme_signing_region(
+    endpoint_config: &AuthSchemeEndpointConfig<'_>,
+) -> Result<Option<SigningRegion>, SigV4SigningError> {
+    use SigV4SigningError::BadTypeInEndpointAuthSchemeConfig as UnexpectedType;
+
+    match extract_field_from_endpoint_config("signingRegion", endpoint_config) {
+        Some(Document::String(s)) => Ok(Some(SigningRegion::from(Region::new(s.clone())))),
+        None => Ok(None),
+        _ => Err(UnexpectedType("signingRegion")),
+    }
+}
+
+#[cfg(feature = "sigv4a")]
+fn extract_endpoint_auth_scheme_signing_region_set(
```

Comment:
would probably recommend moving this into the sigv4a module so that we can gate entirely on a module level